accepted/system-range/system-range.md

```diff
@@ -0,0 +1,312 @@
+//
+// When the upper bound is omitted, it's interpreted to be the length of the
+// receiving collection.
+//
+// Asking for Length or enumerating an unbound range will throw
+// NotSupportedException. The consumer of the range is expected to 
+// bind it for a particular length:
+
+void Print(int[] data, Range range)
+{
+    if (range.IsUnbound)
+        range = range.Bind(data.Length);
+
+    foreach (var index in range)
+        Console.WriteLine(data[index]);
+}
+
+var data = new [] { 'a', 'b', 'c' };
+Print(data, 2..); // Prints c
+Print(data, ..2); // Prints a, b
+Print(data, ..);  // Prints a, b, c
+```
+
+### Negative bounds
+
+```csharp
+// You can also use negative bounds. They are interpreted as being relative
+// to the length of the receiving collection, so -1 is the last element,
+// -2 is the second to last element and so on.
+
+var data = new [] { 'a', 'b', 'c' };
+Print(data, -2.:2);  // Prints b, c
+Print(data, -1..);   // Prints c
+Print(data, -3..-1); // Prints a, b
+```
+
+### Using ranges in APIs
+
+Ranges provide a concise way to provide an `index` and a `length`. This allows
+for making common operations more compact and less confusing at the call side.
+For example, consider this version of the indexer for `String`:
+
+```csharp
+partial class String
+{
+    public string this[Range range] => Substring(range.Index, range.Length);
+}
+```
+
+This allows creating substrings simply by using the indexer:
+
+```csharp
+var helloWorld = "Hello, World!";
+var hello = helloWorld[..5];
+var world = helloWorld[7.:5];
+```
+
+While this might appear to just be minor syntactic sugar it allows for creating
+powerful APIs, such as the indexer for `Tensor<T>` (which you can think of as
+multidimensional array where the dimensions aren't statically known).
+
+```csharp
+partial class Tensor<T>
+{
+    public Tensor<T> this[params Range[] ranges] { get; }
+}
+```
+
+The indexer takes a list of ranges that can be used to get slices of the tensor.
+Imagine you have a two dimensional tensor (in other words a table). Imagine it
+represents test data where each row is a test case and the columns are
+representing the inputs with the last column containing the expected result. Now
+let's say you want to split this data into two tensors, one representing the
+inputs and one the outputs.
+
+This might look like this:
+
+```csharp
+Tensor<float> testDataTable = GetTestData();
+int columnCount = testDataTable.GetLength(1);
+
+// Take all rows and all but the last column
+Tensor<float> inputTable = testDataTable[.., ..-1];
+
+// Take all rows and only the last column
+Tensor<float> outputTable = testDataTable[.., -1..]; 
+```
+
+## Requirements
+
+### Goals
+
+* We need a type that both, plays well with existing concepts in the base class
+  library, as well as the intended syntax for C#
+* We need to ensure we can extend the notion of ranges in the base class library
+  in the future (for example, by supporting negative ranges or element types
```

Comment:
Should be updated since we are proposing negative ranges above.

Reply:
Good call